### PR TITLE
ENH: Add tool to lint diffs

### PR DESCRIPTION
**Pull request by person142**
## Reference issue
None

#### What does this implement/fix?
As discussed on the mailing list:

https://mail.python.org/pipermail/scipy-dev/2020-January/023927.html

we can move towards better code style without causing disruption
by adding extra lint checks for diffs. This adds a script that
lints the diff between the tip of the current branch and where it
split off of master.

#### Additional information
Currently only one new check is enabled (whitespace after `,`).
Do we want others?

### CODE REVIEW

**Comment by andyfaff** [comment]
My personal preferences would be for line length (80 chars), spaces around operators ('9 + 9', not '9+9'), correct indentation of continued lines.

**Comment by ev-br** [comment]
+1 for the idea (did not review details though).

> spaces around operators ('9 + 9', not '9+9')

Not this one please. Consider `2*x + y**3`.

**Comment by andyfaff** [comment]
> Not this one please. Consider 2*x + y**3

I agree. There is a setting that allows that, I wouldn't know the number though. It's designed to allow `2*x + y**3` but not `2*x+y**3`.

**Comment by mattip** [comment]
Probably want to add `pycodestyle==2.5.0` to `pyproject.toml`

It would be cool to activate this as part of `runtests.py`. Does it run on code that is changed but not checked in yet (`git diff`)?

Over at NumPy, we recently started thinking about using fstrings. Rather than refactor all the formatting in all the code, we thought it would be better to refactor as we go. Is there a tool that can detect old string formatting?

**Comment by rgommers** [comment]
> Probably want to add `pycodestyle==2.5.0` to `pyproject.toml`

It's not a build requirement, so I don't see the relevance here.



> It would be cool to activate this as part of `runtests.py`. Does it run on code that is changed but not checked in yet (`git diff`)?

TBD, depends on how it will be implemented. Would be useful if it did indeed.

**Comment by rgommers** [comment]
Ah nice, this is really clean and simple, thanks @person142

**Review by rlucas7 — COMMENTED** [review]
@person142 I left a couple comments thinking about how this might be used in the future scanning commits in rev-list greater (or less) than 100. Overall I'm emphatically +1 for this approach.

File: tools/lint_diff.py
Comment (pvanmulbregt) [reply tools/lint_diff.py]: IIUC this is looking back 100 commits on the branch.  Do we have stats on the number of commits per PR? My expectation is that 100 is reasonable, but maybe there have been some PRs with a lot of commits??
And also looking back on 100 commits on master.  I wonder how many PRs branched more than 100 commits ago, and yet don't need a rebase to avoid conflicts? This actually does seem possible given the elapsed time to review and merge PRs.

Comment (person142) [reply tools/lint_diff.py]: > It seems like a reasonable number but it might be that we want to make that adjustable

I don't think we want it to be adjustable-this should be a simple tool that "just works". I bumped it up to 1000, if that's not enough we can increase it further. The only thing it's really trying to avoid is making a set of all 22,000 commits.

**Comment by person142** [comment]
> It would be cool to activate this as part of runtests.py. Does it run on code that is changed but not checked in yet (git diff)?

Hooked it into `runtests` with the `--pep8` flag. It does run on unstaged/staged but not committed changes.

**Comment by person142** [comment]
I think this is ready for review. I imagine that we can bikeshed on which rules to enable indefinitely, so upon further reflection I suggest that we continue that discussion elsewhere. Once we can lint diffs it's very easy to enable new checks anyway.

**Comment by ev-br** [comment]
> Hooked it into runtests with the --pep8 flag. It does run on unstaged/staged but not committed changes.

So what's the recommended workflow then? If previously it were something like 

```
... hack hack hack ...
$ python runtests.py
... tests pass
$ git commit 
$ git push
```

when is the time for `$ python runtests.py --pep8` and on what branch etc?

File: runtests.py
Comment (ilayn) [inline runtests.py]: Shouldn't this be also a `subprocess.run` call?

**Comment by rlucas7** [comment]
> On Feb 2, 2020, at 10:59 AM, Paul van Mulbregt <notifications@github.com> wrote:
> 
> ﻿
> @pvanmulbregt commented on this pull request.
> 
> In tools/lint_diff.py:
> 
> > +    branch_commits = rev_list('HEAD', 100)
> +    master_commits = set(rev_list('master', 100))
> IIUC this is looking back 100 commits on the branch. Do we have stats on the number of commits per PR? My expectation is that 100 is reasonable, but maybe there have been some PRs with a lot of commits??
> And also looking back on 100 commits on master. I wonder how many PRs branched more than 100 commits ago, and yet don't need a rebase to avoid conflicts? This actually does seem possible given the elapsed time to review and merge PRs.
> 

I’ve had pr’s a couple hundred and in one case a couple thousand behind master. The latter case is less common I’d guess but can happen.

I was thinking that if the plan is to stitch into runtests.py then adding an argument for the number of commits to look back with whatever as the sensible default. It sounds like it is used by pep8 flag so what I was thinking is:

Python runtests.py -v pep8 -n 100
For going back 100 (example).

Python runtests.py -v pep8
As example using whatever is the default.

But if we dont want to have that in the options then we will use whatever the default is always, often it will be too long and occasionally it will be too short.

File: tools/lint_diff.py
Comment (rlucas7) [reply tools/lint_diff.py]: > wonder how many PRs branched more than 100 commits ago, and yet don't need a rebase to avoid conflicts?

How would the conflicts impact the linter? Not clear to me. 

>I don't think we want it to be adjustable-this should be a simple tool that "just works".

I agree I had misunderstood what was happening in the code here, looks like the 100 commits should be reasonable on the 'HEAD' branch. If the 100 is too much on the master then the person should rebase-is this the idea/intent?
New committers often have difficulty we using the rebase command. 

>The only thing it's really trying to avoid is making a set of all 22,000 commits.

I agree that 22K is not needed, it does seem like more would be needed on master than on the branch itself to execute without error though.

**Comment by rlucas7** [comment]
> > Hooked it into runtests with the --pep8 flag. It does run on unstaged/staged but not committed changes.
> 
> So what's the recommended workflow then? If previously it were something like
> 
> ```
> ... hack hack hack ...
> $ python runtests.py
> ... tests pass
> $ git commit 
> $ git push
> ```
> 
> when is the time for `$ python runtests.py --pep8` and on what branch etc?

IIUC the workflow would be something 

> ```
 ... hack hack hack ...
$ python runtests.py
 ... tests pass
$ python runtests.py --pep8
... linter passes 
$ git commit 
$ git push
```
I'm not sure though. If the two `runtest.py` calls could be a single call (running tests and linting) would be even better, I'm always forgetting to lint.

**Comment by rlucas7** [comment]
It looks like the capture_output argument was added to subprocess.run after python 3.6 in 3.7

Compare
https://docs.python.org/3.6/library/subprocess.html
With 
https://docs.python.org/3.7/library/subprocess.html

Are we still supporting 3.6?

-Lucas Roberts

**Comment by tylerjereddy** [comment]
> Are we still supporting 3.6?

Yes, SciPy `1.5.0` will be Python `3.6+`.

File: runtests.py
Comment (ev-br) [inline runtests.py]: `tox.ini` or `lint_diff.ini`?

File: tools/lint_diff.py
Comment (ev-br) [inline tools/lint_diff.py]: there is probably an incantation using `git merge-base`?

**Review by ev-br — APPROVED** [review]
Looks good modulo two small questions. Neither is blocking, so maybe be best to merge it in and iterate as needed.

File: tools/lint_diff.py
Comment (person142) [reply tools/lint_diff.py]: I believe they are different based on https://stackoverflow.com/questions/1527234/finding-a-branch-point-with-git#4991675.

File: runtests.py
Comment (person142) [reply runtests.py]: This one is `tox.ini` since it's the `os.system("pycodestyle scipy benchmarks/benchmarks")` line.

**Comment by person142** [comment]
Ok, CI is green (Python 3.6 compat issues fixed). Re the workflow question, yes it is

```
$ python runtests.py
...
# python runtests.py --pep8
...
```

but I believe it has always been that way. Perhaps it makes sense to include `--pep8` as part of the default invocation, but that is I believe orthogonal to the present change.

If the number of commits that we search back looking for the branch point turns out to be too small, then we can simply bump it. Note that including it as an option does not really help because you won't be able to bump it in CI without a PR doing the bumping anyway.

File: runtests.py
Comment (rlucas7) [reply runtests.py]: @ilayn 
RE: 
`         os.system("pycodestyle scipy benchmarks/benchmarks")`

I'm not even sure we need that line?
On a local branch checkout, I commented out the line, and still caught things. Is this line catching the 
top level python files only?  (e.g. `runtests.py`, `setup.py`, etc) seems weird to put something for that separate from everything else. 

Incidentally, removing the line it speeds up execution substantially.

I then removed the W291 ignore from the ini file (on local PR branch) and added a trailing whitespace (probably my most frequent pep8 violation). The error is still caught without the line. 

```
(scipy-dev) Lucass-MacBook:scipy rlucas$ git diff --staged
diff --git a/runtests.py b/runtests.py
index 4f33707dc..16799add2 100755
--- a/runtests.py
+++ b/runtests.py
@@ -65,8 +65,6 @@ except ImportError:  # old Python
     from imp import new_module
 
 ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__)))
-
-
 def main(argv):
     parser = ArgumentParser(usage=__doc__.lstrip())
     parser.add_argument("--verbose", "-v", action="count", default=1,
@@ -130,7 +128,7 @@ def main(argv):
 
     if args.pep8:
         # Lint the source using the configuration in tox.ini.
-        os.system("pycodestyle scipy benchmarks/benchmarks")
+        # os.system("pycodestyle --show-source --show-pep8 scipy benchmarks/benchmarks") 
         # Lint just the diff since branching off of master using a
         # stricter configuration.
         lint_diff = os.path.join(ROOT_DIR, 'tools', 'lint_diff.py')
@@ -475,3 +473,4 @@ def lcov_generate():
 
 if __name__ == "__main__":
     main(argv=sys.argv[1:])
+
diff --git a/tools/lint_diff.ini b/tools/lint_diff.ini
index 3cf2106e5..117c0d6eb 100644
--- a/tools/lint_diff.ini
+++ b/tools/lint_diff.ini
@@ -1,5 +1,5 @@
 [pycodestyle]
 max_line_length = 79
 statistics = True
-ignore = E121,E122,E123,E125,E126,E127,E128,E226,E251,E265,E266,E302,E402,E501,E712,E721,E731,E741,W291,W293,W391,W503,W504
+ignore = E121,E122,E123,E125,E126,E127,E128,E226,E251,E265,E266,E302,E402,E501,E712,E721,E731,E741,W293,W391,W503,W504
 exclude = scipy/__config__.py
(scipy-dev) Lucass-MacBook:scipy rlucas$ python runtests.py --pep8
./runtests.py:131:89: W291 trailing whitespace
1       W291 trailing whitespace
```

Unless there is a reason to keep it I'd suggest we remove the line, timings:
```
(scipy-dev) Lucass-MacBook:scipy rlucas$ time python runtests.py --pep8
./runtests.py:131:89: W291 trailing whitespace
1       W291 trailing whitespace


real	0m0.414s
user	0m0.325s
sys	0m0.086s 
```
now keeping that line in:
```
(scipy-dev) Lucass-MacBook:scipy rlucas$ time python runtests.py --pep8
./runtests.py:131:61: W291 trailing whitespace
1       W291 trailing whitespace


real	0m39.582s
user	0m38.564s
sys	0m0.340s
```

File: tools/lint_diff.py
Comment (rlucas7) [inline tools/lint_diff.py]: if no issues present still might be nice to see a message, something like "No pycodestyle issues found" or something like that, this need not block the PR to go in though.

**Review by rlucas7 — APPROVED** [review]
Neither of my comments are blocking so I'll mark as approved but would be great to understand if we need that `os.system()` call that @ilayn commented on, it really slows down execution of the test runner.

**Comment by rlucas7** [comment]
> Ok, CI is green (Python 3.6 compat issues fixed). Re the workflow question, yes it is
> 
> ```
> $ python runtests.py
> ...
> $ python runtests.py --pep8
> ...
> ```
> 
> but I believe it has always been that way. Perhaps it makes sense to include `--pep8` as part of the default invocation, but that is I believe orthogonal to the present change.

Agreed. 

> If the number of commits that we search back looking for the branch point turns out to be too small, then we can simply bump it. Note that including it as an option does not really help because you won't be able to bump it in CI without a PR doing the bumping anyway.

I understand now. I had previously understood the intent was for developers. What you say definitely makes sense for CI tool-thanks @person142 

Left two comments, both non-blocking and marked as approved.

File: runtests.py
Comment (person142) [reply runtests.py]: The first call is diffing the entire source tree, and the second call is just linting the diff. Perhaps the former is redundant now, but I can't swear to it.

Comment (rlucas7) [reply runtests.py]: @person142 not sure I follow, I don't think:
`os.system("pycodestyle scipy benchmarks/benchmarks")`
does any diff, unless that is what `pycodestyle` does under the hood but that should be opaque to the caller. 

My understanding is that the `tools/lint.py` is the script where the calls are being made to the diffs. 

Did I misunderstand what you meant?

The call:
`os.system("pycodestyle scipy benchmarks/benchmarks")`
 doesn't seem to capture and forward to `stdout` so if you wanted that then it might make sense to switch but it does seem like if that is what you want to do that should reside in the linter file and not in the test harness but that is just my 2 cents.

**Comment by rgommers** [comment]
This got left unmerged, looks good to go - rerunning CI to check if it still passes.

**Comment by person142** [comment]
Rebased... let's see if that helps.

File: tools/lint_diff.py
Comment (rgommers) [reply tools/lint_diff.py:82]: This makes sense when running manually (but then you already know you're running it anyway); I think the same can be said for the `if errors` block though, some header or footer could help. Anyway, this is polish we can do if it's confusing. Let's merge this as is and see if it works without surprises first:)

In CI this gets printed:
```
$ tools/lint_diff.py
The command "tools/lint_diff.py" exited with 0.
```

**Comment by rgommers** [comment]
Let's give this a go! Thanks @person142, all

**Comment by person142** [comment]
🤞 let's hope it works as expected.

**Pull request merged**
